### PR TITLE
[Tensor] Add outplace method for arithmetic ops

### PR DESCRIPTION
**Pull request by zhoonit**
Add outplace ops with already allocated tensor.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

See also #856

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/tensor/tensor.cpp [hunk 1]
```diff
@@ -183,17 +183,26 @@ Tensor::Tensor(
           this->setValue(i, j, k, l, d[i][j][k][l]);
 }
 
-int Tensor::multiply_i(float const &value) {
+Tensor Tensor::multiply(float const &value, Tensor &out) const {
+  CREATE_IF_EMPTY_DIMS(out, getDim());
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:187]: Let's add a note here that the implementation of `multiply` does not depend on `multiply_i`.
Instead, uses a different methodology as `sscal` only works inplace.

File: nntrainer/tensor/tensor.cpp [hunk 2]
```diff
@@ -219,8 +228,27 @@ int Tensor::add_i(float const &value) {
   return ML_ERROR_NONE;
 }
 
+Tensor Tensor::add(Tensor const &m, Tensor &out, float const alpha) const {
+  CREATE_IF_EMPTY_DIMS(out, getDim());
+  scopy(length(), getData(), 1, out.getData(), 1);
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:233]: do we really need the copy? we can do it in a single call to transform.

File: nntrainer/tensor/tensor.h [hunk 3]
```diff
@@ -250,6 +250,14 @@ class Tensor {
    */
   Tensor multiply(float const &value);
 
+  /**
+   * @brief     multiply value element by element
+   * @param[in] value Divisor
```
Comment (kparichay) [inline nntrainer/tensor/tensor.h:255]: divisor -> multiplier

Comment (zhoonit) [reply nntrainer/tensor/tensor.h]: Done

File: nntrainer/tensor/tensor.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:235]: This will be changed in upcoming PR. Thanks :)

File: nntrainer/tensor/tensor.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:189]: Done

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/tensor/tensor.cpp [hunk 4]
```diff
@@ -219,8 +230,27 @@ int Tensor::add_i(float const &value) {
...
 
+Tensor Tensor::add(float const &value, Tensor &out) const {
+  const float *data = getData();
+
+  CREATE_IF_EMPTY_DIMS(out, getDim());
+
+  std::transform(data, data + length(), out.getData(),
+                 std::bind2nd(std::plus<float>(), value));
+
+  return out;
```
Comment (jijoongmoon) [inline nntrainer/tensor/tensor.cpp:251]: so.. if we are using this... the dimensions is going to change. and if it is the same.. then.. we don't need to another buffer?

Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:251]: you mean out dimension? if uninitialized, it is initialized with `this->getDim()`, if dimension is not correct it is UB 🍫 .

 🍫  is intended as the function is in critical path.

Comment (jijoongmoon) [reply nntrainer/tensor/tensor.cpp:251]: On second thought... It generates a quite of bug which is really hard to find.

Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:251]: I think it's good idea. I'll update on #870 since this patch is being depended upon other PR :)

Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:251]: Update done :)

**Pull request merged**